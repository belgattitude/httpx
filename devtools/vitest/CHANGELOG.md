# @httpx/devtools-vitest

## 0.0.2

### Patch Changes

- [#1872](https://github.com/belgattitude/httpx/pull/1872) [`6b5c38e`](https://github.com/belgattitude/httpx/commit/6b5c38eda03d541c62a1cdf9ba298be5b75087e1) Thanks [@belgattitude](https://github.com/belgattitude)! - Updated browserslist baseline for 2025

  For most users there won't be any change.
  Still around 95% on [browserslist](https://browserslist.dev/?q=ZGVmYXVsdHMsIGNocm9tZSA%2BPSA5NiwgZmlyZWZveCA%2BPSAxMDUsIGVkZ2UgPj0gMTEzLCBzYWZhcmkgPj0gMTUsIGlvcyA%2BPSAxNSwgb3BlcmEgPj0gMTAzLCBub3QgZGVhZA%3D%3D).

  ```
  defaults
  chrome >= 96
  firefox >= 105
  edge >= 113
  safari >= 15
  ios >= 15
  opera >= 103
  not dead
  ```
