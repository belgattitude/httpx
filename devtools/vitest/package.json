{
  "name": "@httpx/devtools-vitest",
  "description": "Shareable dev setups for vitest",
  "version": "0.0.2",
  "private": true,
  "license": "MIT",
  "author": {
    "name": "Vanvelthem SÃ©bastien",
    "url": "https://github.com/belgattitude"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/belgattitude/httpx.git",
    "directory": "devtools/vitest"
  },
  "sideEffects": false,
  "exports": {
    ".": "./src/index.ts",
    "./package.json": "./package.json",
    "./wrangler.toml": "./wrangler.toml"
  },
  "scripts": {
    "clean": "rimraf ./dist ./build ./coverage ./_release",
    "fix-staged": "lint-staged --allow-empty",
    "lint": "eslint . --ext .ts,.tsx,.js,.jsx,.mjs,.cjs,.mts,.cts",
    "typecheck": "tsc --project tsconfig.json --noEmit",
    "install-playwright": "playwright install chromium"
  },
  "peerDependencies": {
    "vitest": "^3.0.8"
  },
  "dependencies": {
    "@belgattitude/eslint-config-bases": "8.3.0",
    "@cloudflare/vitest-pool-workers": "0.9.14",
    "@codspeed/vitest-plugin": "5.0.1",
    "@edge-runtime/vm": "5.0.0",
    "@testing-library/dom": "10.4.1",
    "@vitest/browser": "3.2.4",
    "@vitest/coverage-istanbul": "3.2.4",
    "@vitest/runner": "3.2.4",
    "@vitest/snapshot": "3.2.4",
    "@vitest/ui": "3.2.4",
    "esbuild": "0.25.11",
    "eslint": "8.57.1",
    "playwright": "1.56.1",
    "prettier": "3.6.2",
    "typescript": "5.9.3",
    "vite": "7.1.9",
    "vite-tsconfig-paths": "5.1.4",
    "vitest-browser-react": "1.0.1"
  },
  "devDependencies": {
    "@types/node": "24.8.1",
    "@types/react": "19.2.2",
    "@types/react-dom": "19.2.2",
    "is-ci": "4.1.0",
    "react": "19.2.0",
    "react-dom": "19.2.0",
    "vitest": "3.2.4"
  },
  "engines": {
    "node": ">=18"
  }
}
