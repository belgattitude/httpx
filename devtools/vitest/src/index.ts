export { vitestBenchOptionsConfig } from './vitest-bench-options.config';
