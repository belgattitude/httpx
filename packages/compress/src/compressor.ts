import { uint8ArrayToBase64 } from 'uint8array-extras';

import type { SupportedCompressionAlgorithm } from './compression-algorithm';
import { globalCache } from './global-cache';
import {
  type EncodeStringOptions,
  supportedStringEncodings,
} from './string-encodings';

export class Compressor {
  #algorithm: SupportedCompressionAlgorithm;

  /**
   * Create a new Compressor instance.
   *
   * ```typescript
   * import { Compressor } from '@httpx/compress';
   *
   * const compressor = new Compressor('gzip'); // or 'deflate'
   *
   * const binary = await compressor.toUint8Array('Hello, World! 🦆');
   * const textEncoded = await compressor.toEncodedString('Hello, World! 🦆');
   * ```
   */
  constructor(compressionMethod: SupportedCompressionAlgorithm) {
    this.#algorithm = compressionMethod;
  }

  /**
   * Compress the given data and return it as a Uint8Array binary format
   *
   * ```typescript
   * import { Compressor } from '@httpx/compress';
   *
   * const compressor = new Compressor('gzip');
   * const longString = 'Hello, World! 🦆'.repeat(500_000);
   * const compressedBinary = await compressor.Uint8Array(longString);
   * ```
   * @throws Error
   */
  toUint8Array = async <T extends string | Uint8Array>(
    data: T
  ): Promise<Uint8Array> => {
    const readableStream = new ReadableStream({
      start(controller) {
        controller.enqueue(
          typeof data === 'string'
            ? globalCache.utf8TextEncoder.encode(data)
            : data
        );
        controller.close();
      },
    });
    const compressedStream = readableStream.pipeThrough(
      new CompressionStream(this.#algorithm)
    );

    return new Uint8Array(await new Response(compressedStream).arrayBuffer());
  };

  /**
   * Compress the given data and return it as a base64 encoded string.
   *
   * ```typescript
   * import { Compressor } from '@httpx/compress';
   *
   * const compressor = new Compressor('gzip');
   * const longString = 'Hello, World! 🦆'.repeat(500_000);
   *
   * const compressedString = await compressor.toEncodedString(longString, {
   *   // Default is 'base64'
   *   encoding: 'base64',
   * });
   * ```
   *
   * @throws TypeError if the encoding is not supported
   */
  toEncodedString = async <T extends string | Uint8Array>(
    data: T,
    options?: EncodeStringOptions
  ): Promise<string> => {
    const { encoding = 'base64' } = options ?? {};
    if (!supportedStringEncodings.includes(encoding)) {
      throw new TypeError(`Unsupported string encoding: ${encoding}`);
    }
    return uint8ArrayToBase64(await this.toUint8Array(data), {
      urlSafe: encoding === 'base64_urlsafe',
    });
  };
}
