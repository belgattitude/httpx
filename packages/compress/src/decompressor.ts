import { base64ToUint8Array } from 'uint8array-extras';

import type { SupportedCompressionAlgorithm } from './compression-algorithm';

export class Decompressor {
  #algorithm: SupportedCompressionAlgorithm;

  constructor(algorithm: SupportedCompressionAlgorithm) {
    this.#algorithm = algorithm;
  }

  /**
   * @throws Error
   */
  fromUint8Array = async (data: Uint8Array): Promise<Uint8Array> => {
    const inputReadableStream = new ReadableStream({
      start(controller) {
        controller.enqueue(data);
        controller.close();
      },
    });

    const decompressedStream = inputReadableStream.pipeThrough(
      new DecompressionStream(this.#algorithm)
    );

    const buffer = await new Response(decompressedStream).arrayBuffer();
    return new Uint8Array(buffer);
  };

  /**
   *
   * @throws Error
   */
  fromEncodedString = async (data: string): Promise<string> => {
    return new TextDecoder().decode(
      await this.fromUint8Array(base64ToUint8Array(data))
    );
  };
}
