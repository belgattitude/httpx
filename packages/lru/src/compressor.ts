export const decompress = async (input: Uint8Array): Promise<string> => {
  // Create a ReadableStream from the input Uint8Array
  const inputReadableStream = new ReadableStream({
    start(controller) {
      controller.enqueue(input);
      controller.close();
    },
  });

  // Decompress the ReadableStream using the gzip algorithm
  const decompressedStream = inputReadableStream.pipeThrough(
    new DecompressionStream('gzip')
  );

  // Convert the decompressed stream to a string
  const decompressedArrayBuffer = await new Response(
    decompressedStream
  ).arrayBuffer();
  const decoder = new TextDecoder();
  return decoder.decode(decompressedArrayBuffer);
};
export const compress = async (inputString: string): Promise<Uint8Array> => {
  const encoder = new TextEncoder();
  // Create a ReadableStream from the input string
  const inputReadableStream = new ReadableStream({
    start(controller) {
      controller.enqueue(encoder.encode(inputString));
      controller.close();
    },
  });

  // Compress the ReadableStream using the gzip algorithm
  const compressedStream = inputReadableStream.pipeThrough(
    new CompressionStream('gzip')
  );

  return new Uint8Array(await new Response(compressedStream).arrayBuffer());
};
