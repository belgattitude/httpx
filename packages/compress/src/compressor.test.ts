import { describe } from 'vitest';

import type { SupportedCompressionAlgorithm } from './compression-algorithm';
import { Compressor } from './compressor';
import { Decompressor } from './decompressor';

const testAlgorithms = [
  ['gzip'],
  ['deflate'],
] as const satisfies SupportedCompressionAlgorithm[][];

describe.each(testAlgorithms)('Tests with algorithm "%s"', (algorithm) => {
  const longString = `ðŸ˜Š-abcdef-Ã©Ã Ã¹-012345`.repeat(100);
  const longStringSize = longString.length;

  describe('toEncodedString', () => {
    it('should compress the long string by at least 15 times', async () => {
      const comp = new Compressor(algorithm);
      const compressed = await comp.toEncodedString(longString, 'base64');
      expect(compressed.length).toBeGreaterThan(50);
      expect(compressed.length).toBeLessThan(longStringSize / 15);
    });
    it('should match snapshot', async () => {
      const comp = new Compressor(algorithm);
      const compressed = await comp.toEncodedString(longString, 'base64');
      expect(compressed).toMatchSnapshot();
    });
  });

  describe('fromEncodedString', async () => {
    it('should decompress a compressed string', async () => {
      const comp = new Compressor(algorithm);
      const compressed = await comp.toEncodedString(longString, 'base64');
      const decomp = new Decompressor(algorithm);
      const decompressed = await decomp.fromEncodedString(compressed);
      expect(decompressed).toStrictEqual(longString);
    });
  });
});
