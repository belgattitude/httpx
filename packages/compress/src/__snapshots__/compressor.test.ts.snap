// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests with algorithm "deflate" > toEncodedString > should match snapshot 1`] = `"eJz7MH9Gl25iUnJKapru4ZWHFxzeqWtgaGRsYvphVGJUYlRiVGJUYlRiVGJUYlRisEoAANJ0Tvk="`;

exports[`Tests with algorithm "gzip" > toEncodedString > should match snapshot 1`] = `"H4sIAAAAAAAAA/swf0aXbmJSckpqmu7hlYcXHN6pa2BoZGxi+mFUYlRiVGJUYlRiVGJUYlRiVGKwSgAAtKgf6sQJAAA="`;
